File: linux/ardium_source/presentation/08_Concurrency/concurrency.m
// Topic: Concurrency (GCD)
// Run a task in background

#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        dispatch_group_t group = dispatch_group_create();
        
        NSLog(@"Main: Starting worker");
        dispatch_group_enter(group);
        
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            NSLog(@"Worker running... ID: 1");
            [NSThread sleepForTimeInterval:1.0];
            NSLog(@"Worker done.");
            dispatch_group_leave(group);
        });
        
        NSLog(@"Main: Waiting for worker");
        dispatch_group_wait(group, DISPATCH_TIME_FOREVER);
        
        NSLog(@"Main: Done");
    }
    return 0;
}

File: linux/ardium_source/presentation/08_Concurrency/concurrency.swift
// Topic: Concurrency (Grand Central Dispatch / Async)
// Run a task in background

import Foundation

let group = DispatchGroup()
group.enter()

print("Main: Starting worker")

DispatchQueue.global().async {
    print("Worker running... ID: 1")
    sleep(1)
    print("Worker done.")
    group.leave()
}

print("Main: Waiting for worker")
group.wait()

print("Main: Done")
